### PR TITLE
Add deletion time api

### PR DESCRIPTION
**Pull request by kmdkuk**
- Since there is a permission problem for runner pods to annotate deletion time to themselves, I added an API to expose their own deletion time instead of annotating the runner pods themselves.
  - The deletion time annotation of the runner pod attached by the `slack agent` will be given priority, otherwise `pod sweeper` will check the API of the deletion time and delete it.
- As a result of the above changes, Runner no longer requires `ServiceAccount`, so `ServiceAccount` has been changed to optional.
- I also moved a lot of the processing done in entrypoint to the runner package and split it into several files.

### CODE REVIEW

File: controllers/podsweep_runner.go [hunk 1]
```diff
@@ -15,28 +19,64 @@ import (
...
 //+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;delete
 
+type RunnerPodClient interface {
+	GetDeletionTime(ip string) (string, error)
+}
+
+type RunnerPodClientImpl struct{}
+
+func (c *RunnerPodClientImpl) GetDeletionTime(ip string) (string, error) {
+	url := fmt.Sprintf("http://%s:8080/deletion_time", ip)
```
Comment (masa213f) [inline controllers/podsweep_runner.go:29]: 8080 port is defined as `RunnerMetricsPort`.

File: controllers/podsweep_runner_test.go [hunk 2]
```diff
@@ -132,10 +202,81 @@ var _ = Describe("PodSweeper runner", func() {
...
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "sample",
+								Image: "sample:latest",
+							},
+						},
+					},
+				},
+				time.Now().Add(24 * time.Hour).UTC().Format(time.RFC3339),
```
Comment (masa213f) [inline controllers/podsweep_runner_test.go:274]: It is a future time.

```suggestion
				time.Now().Add(24 * time.Hour).UTC().Format(time.RFC3339),
```

**Review by masa213f — COMMENTED** [review]
Is it possible to create a `runner` package and move some functions to it?

File: runner/runner.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	"github.com/cybozu-go/github-actions-controller/agent"
+	"github.com/cybozu-go/github-actions-controller/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type Runner interface {
+	Run(ctx context.Context) error
+	DeletionTimeHandler(w http.ResponseWriter, _ *http.Request)
+	MetricsInit()
+}
```
Comment (masa213f) [inline runner/runner.go:21]: Why do you use `interface`?

File: runner/runner.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+type RunnerImpl struct {
+	envs Environments
+
+	deletionTime atomic.Value
+}
+
+func NewRunner() (Runner, error) {
+	r := RunnerImpl{
+		envs:         NewRunnerEnvs(),
+		deletionTime: atomic.Value{},
```
Comment (masa213f) [inline runner/runner.go:32]: Do we need this initialization?

File: runner/runner_enviroments.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	constants "github.com/cybozu-go/github-actions-controller"
+)
+
+type Environments struct {
```
Comment (masa213f) [inline runner/runner_enviroments.go:11]: Why do you publish this type?

File: cmd/entrypoint/cmd/root.go [hunk 6]
```diff
@@ -23,79 +16,24 @@ var config struct {
...
-				return err
-			}
-
-			<-ctx.Done()
-			return nil
-		})
+		r.MetricsInit()
+		well.Go(r.Run)
 
 		metricsMux := http.NewServeMux()
```
[the remark this answers is not in the document]
Comment (masa213f) [reply cmd/entrypoint/cmd/root.go:34]: https://github.com/cybozu-go/github-actions-controller/blob/bdf1afa668cae60ccbaabadaa836b7af104f955e/agent/server.go#L67-L89

File: kindtest/helper_test.go [hunk 7]
```diff
@@ -152,6 +155,19 @@ func getRecretedPods(before, after *corev1.PodList) ([]string, []string) {
 	return delPodNames, addPodNames
 }
 
+func getDeletionTime(runnerNS string, po corev1.Pod) (string, error) {
+	stdout, stderr, err := kubectl(
+		"exec", po.Name,
+		"-n", runnerNS,
```
[the remark this answers is not in the document]
Comment (kmdkuk) [reply kindtest/helper_test.go]: thank you, use it.

File: runner/runner.go (hunk 3 above)
Comment (kmdkuk) [reply runner/runner.go]: There was no particular reason for this, so I changed it to Runner struct.

File: runner/runner_enviroments.go (hunk 5 above)
Comment (kmdkuk) [reply runner/runner_enviroments.go]: Changed it to private type.

File: controllers/podsweep_runner_test.go [hunk 8]
```diff
@@ -154,3 +292,17 @@ var _ = Describe("PodSweeper runner", func() {
...
+
+type RunnerPodClientMock struct {
+	deletionTime string
+}
+
+func (c *RunnerPodClientMock) GetDeletionTime(ip string) (string, error) {
+	return c.deletionTime, nil
+}
+
+func NewRunnerPodClientMock(deletionTime string) RunnerPodClientMock {
```
Comment (masa213f) [inline controllers/podsweep_runner_test.go:304]:
```suggestion
func NewRunnerPodClientMock(deletionTime string) *RunnerPodClientMock {
```

File: controllers/podsweep_runner_test.go [hunk 9]
```diff
@@ -30,12 +31,15 @@ var _ = Describe("PodSweeper runner", func() {
...
 		Expect(err).ToNot(HaveOccurred())
 
-		sweeper := NewPodSweeper(
-			mgr.GetClient(),
-			ctrl.Log.WithName("pod-sweeper"),
-			time.Second,
-			organizationName,
-		)
+		runnerPodClient = NewRunnerPodClientMock("")
+
```
Comment (masa213f) [inline controllers/podsweep_runner_test.go:35]: Do you need a line break here?

File: kindtest/helper_test.go [hunk 10]
```diff
@@ -9,13 +9,16 @@ import (
...
 	"sort"
+	"strings"
 	"text/template"
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-github/v33/github"
 	. "github.com/onsi/gomega"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
+
```
Comment (masa213f) [inline kindtest/helper_test.go:20]: Please delete a blank line.

**Review by masa213f — COMMENTED** [review]
1. Please delete the following resources and update documents.
https://github.com/cybozu-go/github-actions-controller/blob/bc31945544bd31580993f9ea1fe81df951671657/config/runnerpool/manifests.yaml#L11-L42

2. Could you customizable the ServiceAccount for runner pods?
Please update the following code.
https://github.com/cybozu-go/github-actions-controller/blob/bc31945544bd31580993f9ea1fe81df951671657/controllers/runnerpool_controller.go#L180
https://github.com/cybozu-go/github-actions-controller/blob/bc31945544bd31580993f9ea1fe81df951671657/api/v1alpha1/runnerpool_types.go#L44

File: api/v1alpha1/runnerpool_types.go [hunk 11]
```diff
@@ -73,6 +73,11 @@ type RunnerPodTemplateSec struct {
 	// List of volumes that can be mounted by containers belonging to the pod.
 	// +optional
 	Volumes []corev1.Volume `json:"volumes,omitempty"`
+
+	// Name of the service account that the Pod use.
+	// +kubebuilder:default="default"
```
Comment (masa213f) [inline api/v1alpha1/runnerpool_types.go:78]: Please confirm the default value is set.

https://github.com/cybozu-go/github-actions-controller/blob/bc31945544bd31580993f9ea1fe81df951671657/api/v1alpha1/runnerpool_webhook_test.go#L65-L68

File: api/v1alpha1/runnerpool_types.go [hunk 12]
```diff
@@ -152,5 +157,5 @@ func (r *RunnerPool) GetRunnerDeploymentName() string {
 
 // GetRunnerServiceAccountName returns the ServiceAccount name for runners.
 func (r *RunnerPool) GetRunnerServiceAccountName() string {
-	return r.Name
+	return r.Spec.Template.ServiceAccountName
 }
```
Comment (masa213f) [inline api/v1alpha1/runnerpool_types.go:161]: This function is no longer needed. Could you delete it?

**Review by masa213f — COMMENTED** [review]
Please update documents.

File: controllers/podsweep_runner.go [hunk 13]
```diff
@@ -87,6 +89,14 @@ func (r *PodSweeper) run(ctx context.Context) error {
 
 		v, ok := po.Annotations[constants.PodDeletionTimeKey]
 		if !ok {
+			v, err = r.runnerPodClient.GetDeletionTime(po.Status.PodIP)
```
Comment (zoetrope) [inline controllers/podsweep_runner.go:92]: I think you can pass context to GetDeletionTime().

File: runner/runner_client.go [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+type Client interface {
+	GetDeletionTime(ip string) (string, error)
+}
+
+type ClientImpl struct{}
+
+func (c *ClientImpl) GetDeletionTime(ip string) (string, error) {
+	url := fmt.Sprintf("http://%s:%d/%s", ip, constants.RunnerMetricsPort, constants.DeletionTimeEndpoint)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (zoetrope) [inline runner/runner_client.go:23]: Please use a passed ctx instead of context.Background().

File: controllers/podsweep_runner.go [hunk 15]
```diff
@@ -87,6 +89,14 @@ func (r *PodSweeper) run(ctx context.Context) error {
 
 		v, ok := po.Annotations[constants.PodDeletionTimeKey]
```
Comment (zoetrope) [inline controllers/podsweep_runner.go:90]: Why are you still using annotations?

Comment (kmdkuk) [reply controllers/podsweep_runner.go:90]: The deletion time annotation is still in use.
When a user requests to extend a failed pod from Slack, the slack-agent will give an annotation.
So pod-sweeper still prioritizes annotation, and if there is no annotation, pod-sweeper uses the API.

Comment (zoetrope) [reply controllers/podsweep_runner.go:90]: Is it not possible for slack-agent to set the deletion time through the API?
It is better not to use multiple means to achieve one purpose.

Comment (kmdkuk) [reply controllers/podsweep_runner.go]: Indeed, you are right.
I'll think about a fix to update the delete time from slack-agent via API.

Comment (kmdkuk) [reply controllers/podsweep_runner.go]: I will deal with this matter in another PR.

File: runner/runner.go [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+func (r *Runner) Run(ctx context.Context) error {
+	registry := prometheus.DefaultRegisterer
+	metrics.Init(registry, r.envs.runnerPoolName)
+
+	env := well.NewEnvironment(ctx)
+	env.Go(r.runListener)
+
+	metricsMux := http.NewServeMux()
+	metricsMux.Handle("/metrics", promhttp.Handler())
+	metricsMux.Handle("/"+constants.DeletionTimeEndpoint, http.HandlerFunc(r.deletionTimeHandler))
```
Comment (zoetrope) [inline runner/runner.go:52]: I think `metricsMux` is not an appropriate name.

**Review by zoetrope — APPROVED** [review]
LGTM

**Pull request merged**